src/heap_ptmalloc_2_27.cpp

```diff
@@ -1813,15 +1813,14 @@ static address_t search_chunk(struct ca_heap* heap, address_t addr)
 static bool fill_heap_block(struct ca_heap* heap, address_t addr, struct heap_block* blk)
 {
 	address_t chunk_addr;
-	struct malloc_chunk achunk;
+	struct malloc_chunk_s achunk;
 	size_t chunksz;
 	size_t size_t_sz = sizeof(INTERNAL_SIZE_T);
-	size_t mchunk_sz = sizeof(struct malloc_chunk);
 
 	// For mmap heap, there is only one block and in-use
 	if (heap->mArena->mType == ENUM_HEAP_MMAP_BLOCK)
 	{
-		if (!read_memory_wrapper(NULL, heap->mStartAddr - size_t_sz, &achunk, mchunk_sz))
+		if (!read_memory_wrapper(NULL, heap->mStartAddr - size_t_sz, &achunk, sizeof(achunk)))
```

Comment:
`read_memory_wrapper` seems always to have a sizeof parameter and the size is always equal to the type to read, so maybe we can add a wrapper function wrapping the `read_memory_wrapper`, so we don't need to pass the sizeof parameter. The function looks like
```c++
template<typename T>
bool read_variable(struct ca_segment* segment, address_t addr, T* val_pointer)
{
       return read_memory_wrapper(segment, addr, val_pointer, sizeof(T));
}
```
With this function, we don't need to pass the size of the parameter anymore.

And most of the time the first argument `segment` is nullptr, we can reorder the template function above to be
```c++
template<typename T>
bool read_variable(address_t addr, T* val_pointer, struct ca_segment* segment =nullptr)
{
       return read_memory_wrapper(segment, addr, val_pointer, sizeof(T));
}
```
And now when reading value, we can call in the following way,
```c++
read_variable(heap->mStartAddr - size_t_sz, achunk)
```
Only two parameters are needed.

Reply:
We could add a convenience function with only two params. The current function is intended to be explicit and requires the caller to think clearly for the accuracy and performance. The variable can be a pointer to the real buffer instead of a local object. The segment parameter is a clue to where the address may be. Otherwise, the memory reader needs to search all segments first. As a matter of fact, we cache the segment of the previous read, which is tried first if the next memory read doesn't provide a segment.

Reply:
Thanks for the explanation. `read_variable` is a supplement to the `read_memory_wrapper` instead of the replacement. I'm drafting a PR in https://github.com/yanqi27/core_analyzer/pull/58